tdrs-backend/tdpservice/parsers/test/test_parse.py

```diff
@@ -1290,3 +1292,38 @@ def test_parse_tribal_section_4_file(tribal_section_4_file):
 
     assert first.FAMILIES_MONTH == 274
     assert sixth.FAMILIES_MONTH == 499
+
+
+@pytest.mark.django_db
+def test_bulk_create_returns_rollback_response_on_bulk_index_exception(test_datafile, mocker):
+    """Test bulk_create_records returns (False, [unsaved_records]) on BulkIndexException."""
+    mocker.patch(
+        'tdpservice.search_indexes.documents.tanf.TANF_T1DataSubmissionDocument.update',
+        side_effect=BulkIndexError('indexing exception')
+    )
+
+    # create some records, don't save them
+    records = {
+        documents.tanf.TANF_T1DataSubmissionDocument: [TANF_T1()],
+        documents.tanf.TANF_T2DataSubmissionDocument: [TANF_T2()],
+        documents.tanf.TANF_T3DataSubmissionDocument: [TANF_T3()]
+    }
+
+    all_created, unsaved_records = parse.bulk_create_records(
+        records,
+        line_number=1,
+        header_count=1,
+        datafile=test_datafile,
+        flush=True
+    )
+
+    assert LogEntry.objects.all().count() == 1
+
+    log = LogEntry.objects.get()
+    assert log.change_message == "Encountered error while indexing datafile documents: indexing exception"
+
+    assert all_created is False
+    assert len(unsaved_records.items()) == 3
+    assert TANF_T1.objects.all().count() == 1
```

Comment:
Two clarification questions. We have one T1 record actually created because we call bulk_create explicitly versus through the whole parsing process. Is the reason the other records aren't created because of the mock? 

Because this isn't happening via the traditional parsing integration tests, we don't call rollback_records (to avoid that infinite recursion with the mock) which would delete the T1/T2/T3 records correct? Which leaves the T1 record alive but not the other two because they don't/shouldn't exist pending the answer to my first Q?

Reply:
I think you've got it - the mock interferes with the `document.update` call, which happens after `bulk_create` for the T1 record. So we manage to create the first record before the exception occurs, then return a response indicating the created records need to be rolled back (which isn't called explicitly by the function, but by the calling function using the return value). If we changed the document that fires the exception to `TANF_T3DataSubmissionDocument`, then all the records would be created.

Reply:
Ok just checking. My weirdness radar was going off when I saw three `unsaved_records` but one T1 object in existence.